### PR TITLE
Add docs, particularly infrastructure

### PR DESCRIPTION
**Pull request by eteq**
This PR adds a first draft of the infrastructure/design document for jdaviz.  It actually renders pretty close to correct directly in github - see https://github.com/eteq/jdaviz/blob/37b64059ba80031a18c0fcb5e7a49ba27533da54/docs/dev/infrastructure.rst .  But if you want to see it as-rendered by sphinx, see also https://eteq.github.io/jdaviz/dev/infrastructure.html .

Note this also includes initial setup for sphinx and some minor package structuring to get sphinx to build correctly.  But the main part of this PR is the infrastructure doc.

cc @brechmos-stsci @astrofrog @nmearl @hcferguson @camipacifici for thoughts/suggestions/comments

### CODE REVIEW

**Comment by brechmos** [comment]
Two initial comments.... 

1) We had talked that the Application Engine part would be in glue-jupyter but the figure doesn't really show that. I think that is an important thing to document - whether it stays in glue-jupyter or if you have talked to others and think it is better in jdaviz. Either way...

2) You might want to add to the figure and text that  there will be a component widget sub-class in jdaviz as well (with similar names) that will implement jdaviz specific functionality.

**Comment by nmearl** [comment]
I think the term "Application Engine" is overloaded in this case. It's both described as being the thing that composes the applications _and_ as the data state management system. I'd argue that the application engine's responsibility is to take the application configuration file (which contains both the layout description _and_ the connections between widgets within the layout) and compose the application (and should exist in glue-jupyter), while the data state management is simply just glue-core.

There also seems to be a delineation based on repositories on the left, which suggests independent concerns for each section; and dependencies on the right, with each "level" being dependent on the one below it. That is, all applications are component widgets + application configurations, but it's the application configurations that exist in jdaviz, and component widgets that live in glue-jupyter. It also suggests that glue-jupyter is dependent on the low-level libraries, which isn't the case, as the low-level library functionality would be implemented as plugins at the application level using the glue-jupyter framework. So, things in the "applications" section would be component widgets (glue-jupyter) + application configurations (jdaviz) + application plugins (jdaviz).

> You might want to add to the figure and text that there will be a component widget sub-class in jdaviz as well (with similar names) that will implement jdaviz specific functionality.

Is this really necessary? Having custom component widget implements? I feel like this is most definitely a use case for composition over inheritance -- there shouldn't be any functionality that can't be composed via the application engine. Otherwise, we're going to develop toward our custom subclass implementations and leave users who wish to compose their own "applications", e.g. within a jupyter notebooks, with less functionally useful component widgets.

**Comment by eteq** [comment]
@brechmos re: 
> We had talked that the Application Engine part would be in glue-jupyter but the figure doesn't really show that. I think that is an important thing to document - whether it stays in glue-jupyter or if you have talked to others and think it is better in jdaviz. Either way...

I got from our discussion that the answer might have to be it's a mix - some of the configuration stuff may need to be jdaviz specific.  So really the dotted line should sort of go right through the application engine?  Or do you mean you think the *engine* should be in `glue-jupyter` but the jdaviz-specific parts should be sort of a separate box?

**Comment by eteq** [comment]
@nmearl -

> I think the term "Application Engine" is overloaded in this case. It's both described as being the thing that composes the applications and as the data state management system. I'd argue that the application engine's responsibility is to take the application configuration file (which contains both the layout description and the connections between widgets within the layout) and compose the application (and should exist in glue-jupyter), while the data state management is simply just glue-core.

I see your point here, but when I think how this should actually be implemented, it's a non-starter to require users to think separately about the data management and the layout/etc.  The user interaction (in e.g. the notebook) should be with a unified thing along the lines of how glupyter works right now with the `jglue` call.  *that* is what I'm thinking of when I talk about the application engine. But perhaps it's useful to separately show the two pieces you are discussing as "sub boxes" inside the engine? and perhaps that also gives a way to address @brechmos's comment by putting the dashed line "through the middle" of the engine?

> There also seems to be a delineation based on repositories on the left ...

Can you offer a more concrete suggestion to address this?  I did *not* intend for this to imply a linear dependency chain, nor a complete separation of concerns. It's more meant to capture all of the *pieces* in a relatively simple set of logical groupings. Basically my thinking was that the goal here is to give people a simple diagram to start with but one where to truly get the details they have to start looking at the details, which would be in *other* parts of the documentation here (once some of it is *implemented*, that is.)

> Is this really necessary? Having custom component widget implements? I feel like this is most definitely a use case for composition over inheritance -- there shouldn't be any functionality that can't be composed via the application engine. 

I share the concern that we want to "push down" as much as possible into glue-jupyter.  But I think this might be one of those practicality over purity cases.  Perhaps we should start with the idea that we want to try to go the "no jdaviz-specific component widgets" direction, but with an understanding (perhaps explicit?) that we might jettison that if it's getting to be really awkward? (unless you see a specific reason we need to do it *right away*, @brechmos ?)

**Comment by hcferguson** [comment]
@eteq et al. Whether in this document or another, I think it is essential to define the interfaces between the different software components that have to work together. The document so far is way too vague on this. If this information isn't going to go in this document, then we need another document sooner rather than later.

At the most abstract level, that means *what* data is getting passed back and forth (e.g. events, bitmapped data downsampled for screen resolution, .... all the way to containers (e.g. Spectrum1D or NDData) with associated uncertainties and physical units). Having identified these interfaces, at this stage we could then highlight which are reasonably mature and which ones need attention. For the ones that are mature, it should be possible in this or an other document to reference the data model and/or API. For the ones that are not, that is a work item that needs to get high priority in the backlog (hence the purpose of trying to write this down). 

Another thing that is distinct but related to the interfaces is where various aspects of *state* are being tracked. If state information is needed in more than one software component, then it needs to be part of the interface. If not, it's nicely isolated. I am quite queasy about this, but that's probably because I just don't have enough information. Hence wanting to see it written down in one place.

**Comment by nmearl** [comment]
> I see your point here, but when I think how this should actually be implemented, it's a non-starter to require users to think separately about the data management and the layout/etc. 

Ah, this sounds like a semantics problem. I don't think the user should be interacting with the application engine _at all_. It should only be used to _compose_ the application, which should then expose an api for interacting with the data state. That is, a user should interact with the one of the four applications you have shown, through an api that is connected to the `DataCollection` or function registry. In my understanding, the application engine is merely a way to ingest some descriptive configuration file and generate an application. In that sense, the user shouldn't even know it exists. E.g. in a notebook context

```python
from specutils import Spectrum1D
import specviz

spec = Spectrum1D.load("/my/data")

# `specviz` is just a wrapper around a set of glue-jupyter start up actions that compose the ui based on a configuration file passed to the application engine, and returns a ui "application"
sv_app = specviz.start()

# Create a standalone profile viewer with, perhaps, a toolbar extended with specviz-specific view/interact options implemented through some glue-jupyter interface. The viewer itself should be dissociated from the main glue-jupyter ui framework, but _not_ from the glue-jupyter state management implemented in the `specviz.start` application.
sv_app.profile_viewer()

# Add spectrum to the specviz for plotting and interacting
sv_app.add_data(spec)

# Get the spectrum from a selected region
spec_subset = sv_app.data_from_subset(sv_app.subsets[0])

...
```

In my mind (and clearly this is hotly debated), the associations go something like:

jdaviz_app_schematic

With the caveat that the scientific libraries implemented in the "Plugins/Function Interface Registry" are for use when being expose through the UI (e.g. a toolbar action, etc). They are then _referenced_ in the declarative configuration file.

> Can you offer a more concrete suggestion to address this?

Depending on the level of simplicity you're going for, perhaps this is the best we can do to understand that we're "building bigger things from smaller things". I thought I could re-arrange and make it work better, but the above diagram (arguably the result of that exercise) does not really make itself clear to non-developers wtf is going on.

> Perhaps we should start with the idea that we want to try to go the "no jdaviz-specific component widgets" direction, but with an understanding (perhaps explicit?) that we might jettison that if it's getting to be really awkward?

I think this is great -- my diagram above is certainly the "purity" case wherein the `specviz` repository just contains the plugins we'd wish to be registry with glue-jupyter (i.e. actions that can be triggered through ui events), and the declaration file which is a description of how the ui is to be composed.

**Comment by nmearl** [comment]
This PR is going to represent the first-draft, and we will merge soon and open another PR with the advances in our thinking on the subject. But some of the content with likely move to glue-jupyter repo.

**Comment by eteq** [comment]
I've updated the text following both the discussion in this thread and several out-of-band in-person follow-ons.  As @nmearl pointed out this is a first draft. There will be a follow-on which tries to channel the discussion in https://github.com/spacetelescope/jdaviz/pull/36#issuecomment-531294522 into something more useful for what @hcferguson and @nmearl have in mind, but inside *glue-jupyter* instead of here, and we will link this document to that one when it's ready.

File: docs/dev/infrastructure.rst [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+
+Application Engine
+------------------
+
+The applications are driven by a shared layer that connects the "high-level" layers to the "low-level" layers discussed below.  The application engine manages this.  Or more concretely, the application engine is the Python-level object that can be "shown" in any of the interfaces to actually interact with a particular application.  It contains several sub-pieces to acheive this goal.  The most directly-used portion of this is the "layout configuration" management - `jdaviz` applications specify their UI layout they use via this part of the application engine. The application engine then actually performs the layout using glue-jupyter_, `ipywidgets`, and other layout javascript/python libraries like ipyvuetify_.  The application engine also manages events and data, primarily via the `glue-core <https://github.com/glue-viz/glue>`_ library and its built-in functionality for both. In this sense the application engine also provides the interface for registering new UI functionality (although this happens indrectly via glue).
```
Comment (nmearl) [inline docs/dev/infrastructure.rst]:
```suggestion
The applications are driven by a shared layer that connects the "high-level" layers to the "low-level" layers discussed below.  The application engine manages this shared layer.  Or more concretely, the application engine is the Python-level object that can be accessed by the user in any of the interfaces to interact with a particular application.  It contains several sub-pieces to achieve this goal.  The most directly-used portion of this is the layout configuration management - `jdaviz` applications specify the UI layout they use via this part of the application engine. The application engine then constructs the layout using glue-jupyter_, `ipywidgets`, and other layout javascript/python libraries like ipyvuetify_. The application engine also leverages the built-in functionality of the `glue-core <https://github.com/glue-viz/glue>`_ library in managing application-level events and data. In this sense the application engine also provides the interface for registering new UI functionality (although this happens indirectly via glue). 
```

Some notes. I do find this _really_ vague, but that's probably unavoidable. I'm not quite sure what you were getting at in some cases, so I tried to be a bit more specified with how I interpreted your intent. I'm also not sure if the last sentence is supposed to be swapped with the second to last sentence?

File: docs/dev/infrastructure.rst [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+
+Component Widgets
+=================
+
+The "component widget" layer is the first of the "low-level" layers - i.e., the layers that actual implement specific visualization and analysis functionality. These widgets are self-contained and while they are meant to be composed in applications, for the notebook interface component widgets can and should be used directly by users for specialized scientific workflows.  Component widgets in principle can be developed in any framework that can be exposed as an `ipywidgets` widget, although currently the plan is that most will be glue-jupyter_ viewers combined with ipyvuetify_ layouts. The general goal is to push any functionality necessary for these widgets upstream into glue-jupyter_ and not confine them to `jdaviz`, but with allowances that some customization may need to be necessary for `jdaviz`/JWST-specific elements.
```
Comment (nmearl) [inline docs/dev/infrastructure.rst]:
```suggestion
The "component widget" layer is the first of the "low-level" layers - i.e., the layers that actual implement specific visualization and analysis functionality. These widgets are self-contained and in general are meant to be composed in applications. However, for the notebook interface, component widgets can and should be used directly by users for specialized scientific workflows.  Component widgets in principle can be developed in any framework that can be exposed as an `ipywidgets` widget, although currently the plan is that most will be glue-jupyter_ viewers combined with ipyvuetify_ layouts. The general goal is to push any functionality necessary for these widgets upstream into glue-jupyter_ and not confine them to `jdaviz`, but with allowances that some customization may need to be necessary for `jdaviz`/JWST-specific elements.
```

File: docs/dev/infrastructure.rst (hunk 1 above)
Comment (eteq) [reply docs/dev/infrastructure.rst]: I hadn't thought about it but I guess it does seem better swapped now that you mention it.

**Pull request merged**